src/main/java/io/github/genomicdatainfrastructure/daam/services/CreateApplicationService.java

```diff
@@ -50,33 +47,30 @@ public void createRemsApplication(CreateApplication createApplication, String us
     }
 
     private List<Long> getCatalogueItemIds(List<String> datasetIds, String userId) {
-        List<CompletableFuture<Long>> futures = datasetIds.stream()
+        var futures = datasetIds.stream()
                 .map(datasetId -> fetchCatalogueItemIdAsync(userId, datasetId))
                 .toList();
 
-        List<Long> result;
         try {
-            result = futures.stream()
+            return futures.stream()
                     .map(CompletableFuture::join)
-                    .collect(Collectors.toList());
+                    .toList();
         } catch (CompletionException e) {
-            if (e.getCause() instanceof CatalogueItemNotFoundException) {
-                throw (CatalogueItemNotFoundException) e.getCause();
+            if (e.getCause() instanceof RuntimeException runtimeException) {
+                throw runtimeException;
             }
 
-            throw e;
+            throw new RuntimeException(e.getCause());
```

Comment:
if you are not using CatalogueItemNotFoundException, then remove it maybe? i was also mapping it here

https://github.com/GenomicDataInfrastructure/gdi-userportal-access-management-service/blob/main/src/main/openapi/daam.yaml#L199-L203

Reply:
The thing is i think with new RuntimeException, CatalogueItemNotFoundException will not be catched by mapper here

https://github.com/GenomicDataInfrastructure/gdi-userportal-access-management-service/blob/main/src/main/java/io/github/genomicdatainfrastructure/daam/api/CatalogueItemNotFoundExceptionMapper.java

Reply:
Thanks for asking. `CompletionException` is a wrapper exception, which means it happened inside a `CompletableFuture`. So we needed to unwrap to get the cause - this case `CatalogueItemNotFoundException`, otherwise the mapper would not find it, thus a 500 exception would be throw. 

Usually there are two approaches, either we unwrap everytime and throw the cause instead, and then we code a mapper for each expected exception. This is the approach you started and then I just made it generic.

Or we have a unique `CompletionExceptionMapper` that captures the completion exception's cause and the build the error response for each type.

For readability I prefer the first one.

The remaining point is for checked exceptions. Unfortunetally the first method is less effective, as we need to wrap checked exceptions again to be able to throw without method signature changes.

In my experience though, checked exception are seldomly used.